scipy/stats/_qmc.py

```diff
@@ -306,6 +311,131 @@ def discrepancy(
+
+    See Also
+    --------
+    discrepancy
+
+    Notes
+    -----
+    The discrepancy can serve as a simple measure of quality of a random sample.
+    This measure is based on the geometric properties of the distribution of points
+    in the sample, such as the minimum distance between any pair of points, or
+    the mean edge length in a minimum spanning tree.
+
+    The higher the value is, the better the coverage of the parameter space is.
+    Note that this is different from `scipy.stats.qmc.discrepancy`, where lower
+    values correspond to higher quality of the sample.
+
+    It is possible to calculate two metrics from the minimum spanning tree:
+    the mean edge length and the standard deviation of edges lengths. Using
+    both metrics offers a better picture of uniformity than either metric alone,
+    with higher mean and lower standard deviation being preferable (see [1]_
+    for a brief discussion). This function currently only calculates the mean
+    edge length.
+
+    References
+    ----------
+    .. [1] Franco J. et al. "Minimum Spanning Tree: A new approach to assess the quality
+       of the design of computer experiments." Chemometrics and Intelligent Laboratory
+       Systems, 97 (2), pp. 164-169, 2009.
+
+    Examples
+    --------
+    Calculate the quality of the sample using the minimum euclidean distance
+    (the defaults):
+
+    >>> import numpy as np
+    >>> from scipy.stats import qmc
+    >>> np.random.seed(12345)
+    >>> sample = np.random.random_sample((50, 2))
+    >>> qmc.geometric_discrepancy(sample)
+    0.018894832717486352
+
+    Calculate the quality using the mean edge length in the minimum
+    spanning tree:
+
+    >>> qmc.geometric_discrepancy(sample, method='mst')
+    0.0938161477108559
+
+    Display the minimum spanning tree and the points with
+    the smallest distance:
+
+    >>> import matplotlib.pyplot as plt
+    >>> from matplotlib.lines import Line2D
+    >>> from scipy.sparse.csgraph import minimum_spanning_tree
+    >>> from scipy.spatial.distance import cdist, pdist, squareform
+    >>> dist = pdist(sample)
+    >>> mst = minimum_spanning_tree(squareform(dist))
+    >>> edges = np.where(mst.toarray() > 0)
+    >>> edges = np.array(edges).T
+    >>> min_dist = np.min(dist)
+    >>> min_idx = np.argwhere(squareform(dist) == min_dist)[0]
+    >>> fig, ax = plt.subplots()
+    >>> plt.gca().set_aspect('equal')
+    >>> for edge in edges:
+    ...     ax.plot(sample[edge, 0], sample[edge, 1], c='k')
+    >>> ax.scatter(sample[:, 0], sample[:, 1])
+    >>> ax.add_patch(plt.Circle(sample[min_idx[0]], min_dist, color='red', fill=False))
+    >>> markers = [
+    ...     Line2D([0], [0], marker='o', lw=0, label='Sample points'),
+    ...     Line2D([0], [0], color='k', label='Minimum spanning tree'),
+    ...     Line2D([0], [0], marker='o', lw=0, markerfacecolor='w', markeredgecolor='r',
+    ...            label='Minimum point-to-point distance'),
```

Comment:
Trying to make it fit
```suggestion
    ...     Line2D([0], [0], marker='o', lw=0, label='Sample'),
    ...     Line2D([0], [0], color='k', label='MST'),
    ...     Line2D([0], [0], marker='o', lw=0, markerfacecolor='w', markeredgecolor='r',
    ...            label='mindist'),
```

Reply:
Done, though I find the resulting legend a bit cryptic

Reply:
I am happy to do otherwise if you find a way for the legend to not being cut.

Not sure if you've seen that you can check here the rendered documentation. You just have to click on the job which reads `Check rendered doc...` once it's green/passing.

Also taking the opportunity here to ask you to put these in your commit messages until we have tests: `[skip actions] [skip cirrus]`. This will just let CircleCI run to build the doc. And once we just want the tests to run, you will be able to do `[skip circle]`. Thanks!!

Reply:
That should be done.

My problem was that running `make html` does not seem to regenerate the plots resulting from docstrings. Is there some trick to do it (other than deleting `doc/build` and generating the whole thing from scratch, which takes ages...)?

Reply:
That's strange, but Sphinx... What you can try is to delete the output html page and Sphinx should detect that it has to rebuild it.

Reply:
Unfortunately, deleting the generated `html` or `png` files seems to have no effect - the image seems to be cached somewhere else.

Reply:
mmm ok, and by `png` you tried to remove `doc/build/plot_directive/reference/generated/fig.png` or just what was in the `html` folder?

Reply:
I tried to remove the figure in the `html` directory. Removing it from the `plot_directive` does the trick, thanks!

Once I did this another interesting bit of behaviour was revealed. The code of the docstring for `geometric_discrepancy` contains this line:
```
>>> rng = np.random.default_rng(191468432622931918890291693003068437394)
```
whereas the rendered html becomes:
```
>>> rng = np.random.default_rng()
```
and a different graph is created each time.

I was not aware that something would try to erase the seed:)

Reply:
Yes this is some magic (I actually introduced) to not present a seed in the doc. We don't want users to copy paste the seed value and in the end transform a RNG into a fixed/bad "NG".

Before we used to have a mechanism that would inject the seed into `>>> rng = np.random.default_rng()` but this is really not convenient for developers as you need to first put the seed on your side, run the doctests, get the right numbers and then remove the seed. And also if you would run the doctests manually on your side with an IDE, that would fail.

With the new mechanism, you can in the code put the seed you want. And still, end users reading the doc would not see any seed.

There is a bit more details there https://scipy.github.io/devdocs/dev/contributor/rendering_documentation.html#examples-and-randomness

Reply:
Pleasure 😃

To be clear, the seed is removed from the rendered HTML, but it is still used when building the figure.

Note that I would not mind having a varying figure in that case. Also here an alternative would be to hand pick a design of something like 10-20 points if we really want to tell a story about the points themselves, etc.

Reply:
> To be clear, the seed is removed from the rendered HTML, but it is still used when building the figure.

That's not what I see though. Every time I regenerate it, I get a different figure.

Not a big deal of course, just a bit surprising...